### DISCUSSION TITLE
API Suggestions :D

### DISCUSSION THREAD

**Discussion by exzachlyvv** [SeaQL/sea-orm#294] [Ideas]
Thank you for this ORM! It's hard to imagine a better developer API for this ORM. But here are a few thoughts. 

*Disclaimer:* I come from Laravel, which uses EloquentORM. Most of these suggestions come from what I see working very well over there.

Potential improvements:
1. Naming for the `find_also_related(Fruit)` and `find_with_related(Fruit)` is a little clunky. Consider:
    1. `with(Fruit)` for eager loading
        1. Maybe a with_one() and a with_many() ?
    2. Could we find a way to use a single method for One to One and One to Many? Usually, I'm not concerned with the relationship type. I just want the relation loaded.
2. More coming as I get more familiar!

Another thought:
From the documentation: `Finally, there is no singleton (global context) in SeaORM.` I understand why. All good. However, consider the value of making developers worry about passing the connections around their applications. Serious question, is there **any** value? It sure is a pain in the butt. If there is no value, it is more tempting to try and work through this problem.

I know the singleton pattern has its problems too, especially for Rust. And in a multithreaded webserver context, this is not going to be an easy feature to add. But thought I would start that dialogue and see what we can do :D

**Comment by tyt2y3** [comment]
Hi there, welcome! Thank you for your feedback, we need you. And thank you for actually taking the time to read our docs )

(And thank you for actually make use of the discussions tab ))

Yes, we are definitely heavily inspired by Laravel and it's our goal to make people coming from a dynamic language feel right at home.

> Finally, there is no singleton (global context) in SeaORM

Well, SeaORM is an ORM, not a web framework. Rust is multi-modal, meaning there could be multiple scenario (web, embedded, desktop). We do bundle Rocket framework integration into our codebase, so at least if you are using Rocket, you don't have to bother. 

https://github.com/SeaQL/sea-orm/blob/master/examples/rocket_example/src/main.rs

(You made a good point, I should add one more paragraph to the docs)

> Maybe a with_one() and a with_many()

It's a forward looking strategy for now. It's far easier to have long function names and then introduce shorter ones and deprecate the longer ones some point in the future, than to having a too general function name and having to deal with ambiguous names later. When the feature set is more or less stable, I'd be happy to introduce shorthands for some of them.

> Could we find a way to use a single method for One to One and One to Many? Usually, I'm not concerned with the relationship type. I just want the relation loaded.

Sorry on this one, I don't really understand, do you mean the `has_one` and `has_many` in `RelationDef` ? Right now they don't actually make a difference in query, so probably you can always use `has_many` (indeed, `has_one` only implies there is an additional unique constraint on the foreign key column).

**Reply by tyt2y3** [reply]
https://github.com/SeaQL/seaql.github.io/commit/e1c12c5c618ed510597509d31192f232d05c2ca2

**Comment by exzachlyvv** [comment]
Very thorough response thank you!

`Could we find a way to use a single method for One to One and One to Many?` I was talking about the names `find_also_related()` and `find_with_related()`. I wonder if we would want to share a simple same name for one to one and one to many?
- But I love your point about starting verbose. We can always add the convenience on top as time goes on. :D So disregaurd
